From b75d57af3a78d59d3700e040d220fb78a69ebcd1 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sat, 7 Oct 2017 18:58:31 +0200
Subject: [PATCH 1/1] efi_loader: Exit must remove loaded image handle

The handle associated with the loaded image of an EFI application
has to be removed when the Exit bootservice is called.

Same is true for EFI service images if the exit code is not
EFI_SUCCESS.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_boottime.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/lib/efi_loader/efi_boottime.c b/lib/efi_loader/efi_boottime.c
index 40cb332cc3..56568e371b 100644
--- a/lib/efi_loader/efi_boottime.c
+++ b/lib/efi_loader/efi_boottime.c
@@ -1148,6 +1148,16 @@ void efi_setup_loaded_image(struct efi_loaded_image *info, struct efi_object *ob
 	list_add_tail(&obj->link, &efi_obj_list);
 }
 
+/*
+ * Free all resources linked to the loaded image handle.
+ *
+ * @loaded_image_info:	handle of the loaded image
+ */
+void efi_remove_loaded_image_handle(struct efi_loaded_image *loaded_image_info)
+{
+	list_del(loaded_image_info->link);
+}
+
 /*
  * Load an image using a file path.
  *
@@ -1337,6 +1347,16 @@ static efi_status_t EFIAPI efi_exit(efi_handle_t image_handle,
 	efi_restore_gd();
 
 	loaded_image_info->exit_status = exit_status;
+
+	/*
+	 * If the image is not a driver meant to stay in the system, the
+	 * handle has to be removed.
+	 */
+	if (loaded_image_info->image_code_type == EFI_LOADER_CODE ||
+	    exit_status != EFI_SUCCESS) {
+		efi_remove_loaded_image_handle(loaded_image_info);
+	}
+
 	longjmp(&loaded_image_info->exit_jmp, 1);
 
 	panic("EFI application exited");
-- 
2.14.1

